### ISSUE TITLE
Modal.js and Accordion.js use of getInstanceId causes all snapshots to need update every time new modals are added

### ISSUE THREAD

**Issue by scottdickerson** [carbon-design-system/carbon#6561]
## What package(s) are you using?

- [ ] `carbon-components`
- [x] `carbon-components-react`

## Detailed description

> Describe in detail the issue you're having.

Because unique ids for Modal, Accordion and other components that need instance Ids are dynamically generated based on load order, if we add more Modals to our codebase, downstream snapshots will tag these ids as changed and require a update.  This is confusing to developers as they haven't updated any of those downstream components.  

> Is this issue related to a specific component?
any compnoent using the setupGetInstanceId method. 

Can we please add an optional id prop to each of these components, that if set is always used by the component (instead of the dynamically generated id).  This would allow us to create snapshots in a deterministic fashion that weren't reliant at runtime.

> What did you expect to happen? What happened instead? What would you like to
> see changed?

> What browser are you working in?

> What version of the Carbon Design System are you using?

> What offering/product do you work on? Any pressing ship or release dates we
> should be aware of?

## Steps to reproduce the issue

![image](https://user-images.githubusercontent.com/6663002/88555944-0d44cc00-cfee-11ea-8472-5e6ac93616dc.png)

> Please create a reduced test case in CodeSandbox
>
> - Style and vanilla JS:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/components/examples/codesandbox
> - React:
>   https://codesandbox.io/s/github/carbon-design-system/carbon/tree/master/packages/react/examples/codesandbox

## Additional information

- Screenshots or code
- Notes

**Comment by jannicz** [comment]
@joshblack  I noticed that this issue also causes incompatibilities with server side rendering. I am using **carbon-components-react 7.22** in combination with **Nextjs 9.5**. When including components that make use of [getInstanceId()](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/Modal/Modal.js#L197), following problem occurs:

- The client renders the component on each page reload with the same instance ID
- The server increments the id on each page request

This results in two different versions of the HTML DOM (one on the server, the other on the client), that actually should be the same. Nextjs complains with the following error message:

<img width="843" alt="getInstanceId" src="https://user-images.githubusercontent.com/9930138/96287063-a843b580-0fe1-11eb-946e-a4558c76db5c.png">

I am suggesting the following change to the affected components:

If the id (i.e. `id="modal01"`) is set by the developer as input property, use this custom id for generating the instanceId like so:
`modalInstanceId = 'modal-${this.props.id || getInstanceId()}';`

This would highly improve Carbons compatibility with SSR frameworks.

**Comment by davesteinberg** [comment]
I'll go further: This pattern is harmful, and you should try to remove it. It makes Carbon a minefield for anyone doing rendering on the server and client. That's the future (and, to a large extent, the present) of React. Generating IDs breaks hydration (as @jannicz has pointed out) and who knows what else.

I just spent more than an hour figuring out why my very basic usage of `Tag` was generating "Prop 'id' did not match." errors in Next.js and working around the problem. It's tricky because I don't need, and can't easily create, unique, stable ids for every `Tag` on my page. And my tags aren't filters, so these generated IDs aren't even used for anything.

It's a really frustrating experience for a new user of this library.

**Comment by joshblack** [comment]
@davesteinberg definitely agreed, thankfully in React 18 we have access to the `useId` hook which will help to create stable id's in the SSR use-case.

In the interim, we've tried to move over to a stable `useId` hook for component based on what Reach UI has done. Unfortunately, in components that are still `class`-based we haven't had a chance to update them to this new approach.

I'll see if there is something we can do in the interim for the `class`-based components, I also went and made this issue: https://github.com/carbon-design-system/carbon/issues/11513 for converting to `React.useId` once we support React 18, as well.
